File: csr_matrix_generator.py
#!/usr/bin/env python3
import random
import sys

def main():
    height = 0
    height = 0
    p = 0.
    oneliner = ""
    if len(sys.argv) > 1:
        if len(sys.argv) != 4: #sys.argv[1] == "-h" or sys.argv[1] == "--help" or 
            print(f"You can use params like this: {sys.argv[0]} <height> <width> <proba>")
            exit(0)
        else:
            height = int(sys.argv[1])
            width = int(sys.argv[2])
            p = float(sys.argv[3])
    else:
        height = int(input("Matrix height: "))
        width = int(input("Matrix width: "))
        p = float(input("1-value probability : "))

    # data = [
    #     1 if random.random() < p else 0
    #     for _ in range(height * width)
    # ]
    pointers = [0]
    columns = []
    nnz=0
    for i in range(height):
        col = -1
        while col < width:
            col += random.randint(1, int(1/p))
            if col < width:
                nnz += 1
                columns.append(col)
        pointers.append(nnz)
        
    # print(f"u_int32_t mat_p[{len(pointers)}] = {{" + ", ".join(str(v) for v in pointers) + "};")
    # print(f"u_int32_t mat_j[{len(columns)}] = {{" + ", ".join(str(v) for v in columns) + "};")
    
    print(f"u_int32_t mat_p[{len(pointers)}] = {{{pointers[0]}")
    for v in range(1,len(pointers)):
        print(f", {pointers[v]}")
    print("};")
    
    
    print(f"u_int32_t mat_p[{len(columns)}] = {{{columns[0]}")
    for v in range(1,len(columns)):
        print(f", {columns[v]}")
    print("};")

if __name__ == "__main__":
    main()

File: csr_matrix_generator.cc
#include <iostream>
#include <vector>
#include <cstdint>
#include <cstdlib>
#include <ctime>

int main(int argc, char* argv[]) {
    std::srand(static_cast<unsigned>(std::time(nullptr)));

    uint32_t height = 0;
    uint32_t width = 0;
    double p = 0.0;

    if (argc > 1) {
        if (argc != 4) {
            std::cout << "Usage: " << argv[0] << " <height> <width> <proba>\n";
            return 0;
        }
        height = static_cast<uint32_t>(std::stoi(argv[1]));
        width  = static_cast<uint32_t>(std::stoi(argv[2]));
        p      = std::stod(argv[3]);
    } else {
        std::cout << "Matrix height: ";
        std::cin  >> height;
        std::cout << "Matrix width: ";
        std::cin  >> width;
        std::cout << "1-value probability: ";
        std::cin  >> p;
    }

    std::vector<uint32_t> pointers;
    std::vector<uint32_t> columns;
    pointers.reserve(height + 1);
    pointers.push_back(0);
    columns.reserve(1.5*p*(height*width));

    uint32_t nnz = 0;
    uint32_t maxStep = static_cast<uint32_t>(1.0 / p);

    for (uint32_t i = 0; i < height; ++i) {
        uint32_t col = -1;
        while (col < width - 1 || col == -1) {
            uint32_t step = (std::rand() % maxStep) + 1;
            col += step;
            if (col < width) {
                ++nnz;
                columns.emplace_back(col);
            }
        }
        pointers.emplace_back(nnz);
    }
    std::cerr << "matrix generated.\nWriting it out...\n";

    std::cout << "u_int32_t mat_p[" << pointers.size() << "] = {";
    for (size_t i = 0; i < pointers.size(); ++i) {
        if (i > 0) std::cout << ",";
        std::cout << pointers[i];
    }
    std::cout << "};\n\n";

    std::cout << "u_int32_t mat_j[" << columns.size() << "] = {";
    for (size_t i = 0; i < columns.size(); ++i) {
        if (i > 0) std::cout << ",";
        std::cout << columns[i];
    }
    std::cout << "};\n";

    return 0;
}
